share/spack/setup-env.sh

```diff
@@ -235,28 +227,65 @@ if ! _spack_fn_exists use && ! _spack_fn_exists module; then
 fi;
 
 
+function _spack_finish_vars() {
+    if test $_spack_var_done -eq 0 ; then
+        wait $_spack_var_pid
+        source $_spack_var_file
+        rm -f $_spack_var_file
+        if [ "${_sp_module_prefix}" != "not_installed" ]; then
+            #activate it!
+            export MODULE_PREFIX=${_sp_module_prefix}
+            _spack_pathadd PATH "${MODULE_PREFIX}/Modules/bin"
+        fi
+        _spack_var_done=1
+    fi
+}
+
 #
 # make available environment-modules
 #
 if [ "${need_module}" = "yes" ]; then
-    eval `spack --print-shell-vars sh,modules`
-
-    # _sp_module_prefix is set by spack --print-sh-vars
-    if [ "${_sp_module_prefix}" != "not_installed" ]; then
-        #activate it!
-        export MODULE_PREFIX=${_sp_module_prefix}
-        _spack_pathadd PATH "${MODULE_PREFIX}/Modules/bin"
-        module() { eval `${MODULE_PREFIX}/Modules/bin/modulecmd ${SPACK_SHELL} $*`; }
-    fi;
+    _spack_var_list="sh,modules"
+    _spack_no_mods=1
 else
-    eval `spack --print-shell-vars sh`
-fi;
+    _spack_var_list="sh"
+fi
+
+# Also set by --print-shell-vars, init here to avoid duplication
+_sp_module_prefix="not_installed"
+_spack_var_file=$(mktemp)
+spack --print-shell-vars $_spack_var_list > $_spack_var_file 2> /dev/null &
+_spack_var_pid=$!
+_spack_var_done=0
+if _spack_fn_exists use ; then
+  eval orig_"$(declare -f use)"
+  function use () {
+    _spack_finish_vars
+    _spack_pathadd DK_NODE    "${_sp_dotkit_root%/}/$_sp_sys_type"
+    orig_use "$@"
+  }
+fi
+# _sp_module_prefix is set by spack --print-sh-vars
+if test "${need_module}" = "yes" ; then
+    module() {
+      eval `${MODULE_PREFIX}/Modules/bin/modulecmd ${SHELL:=bash} $*`
+    }
+fi
+if _spack_fn_exists module ; then
+  eval orig_"$(declare -f module)"
+  function module () {
```

Comment:
See above w.r.t. #8570 -- I think it has to call `module`, or at least consider that `modulecmd` may not be findable.

The other worry I have here is that this won't work for users who load modules *after* they log in.  I've seen users at LC manually source the module initialization scripts long after startup.  Now they need to know to source Spack init after module init too.  That doesn't seem reliable.  The current implementation just ensures that `MODULEPATH` is set, so modules will pick up Spack stuff even if loaded later.

Reply:
That's a fair concern.  The main goal is just to make it so that the initial load doesn't have to wait.  Really, having a supported mechanism for outputting a set of environment variables that avoids having to call into python (with the agreement that the info may be stale) might be a better way to do this.  The mechanism to print out the contents is already there, maybe adding a way to tell this script not to re-compute paths would be more palatable?